core/mutation/mutation.go

```diff
@@ -0,0 +1,211 @@
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	tpb "github.com/google/keytransparency/core/proto/keytransparency_v1_types"
+	"github.com/google/trillian"
+)
+
+// Server holds internal state for the monitor server core functionality.
+type Server struct {
+	logID     int64
+	mapID     int64
+	tlog      trillian.TrillianLogClient
+	tmap      trillian.TrillianMapClient
+	mutations mutator.Mutation
+	factory   transaction.Factory
+}
+
+// New creates a new instance of the monitor server.
+func New(logID int64,
+	mapID int64,
+	tlog trillian.TrillianLogClient,
+	tmap trillian.TrillianMapClient,
+	mutations mutator.Mutation,
+	factory transaction.Factory) *Server {
+	return &Server{
+		logID:     logID,
+		mapID:     mapID,
+		tlog:      tlog,
+		tmap:      tmap,
+		mutations: mutations,
+		factory:   factory,
+	}
+}
+
+// GetMutations returns a list of mutations paged by epoch number.
+func (s *Server) GetMutations(ctx context.Context, in *tpb.GetMutationsRequest) (*tpb.GetMutationsResponse, error) {
+	// Get signed map root by revision.
+	resp, err := s.tmap.GetSignedMapRootByRevision(ctx, &trillian.GetSignedMapRootByRevisionRequest{
+		MapId:    s.mapID,
+		Revision: in.Epoch,
+	})
+	if err != nil {
+		log.Printf("GetSignedMapRootByRevision(%v, %v): %v", s.mapID, in.Epoch, err)
+		return nil, grpc.Errorf(codes.Internal, "Get signed map root failed")
+	}
+
+	// Get highest and lowest sequence number.
+	highestSeq := uint64(resp.GetMapRoot().GetMetadata().HighestFullyCompletedSeq)
+	lowestSeq, err := s.lowestSequenceNumber(ctx, in.PageToken, in.Epoch-1)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read mutations from the database.
+	txn, err := s.factory.NewTxn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("NewDBTxn(): %v", err)
+	}
+	maxSequence, mRange, err := s.mutations.ReadRange(txn, lowestSeq, highestSeq, in.PageSize)
+	if err != nil {
+		log.Printf("mutations.ReadRange(%v, %v, %v): %v", lowestSeq, highestSeq, in.PageSize, err)
+		if err := txn.Rollback(); err != nil {
+			log.Printf("Cannot rollback the transaction: %v", err)
+		}
+		return nil, grpc.Errorf(codes.Internal, "Reading mutations range failed")
+	}
+	if err := txn.Commit(); err != nil {
+		return nil, fmt.Errorf("txn.Commit(): %v", err)
+	}
+	indexes := make([][]byte, 0, len(mRange))
+	mutations := make([]*tpb.Mutation, 0, len(mRange))
+	for _, m := range mRange {
+		mutations = append(mutations, &tpb.Mutation{Update: m})
+		indexes = append(indexes, m.GetKeyValue().GetKey())
+	}
+	// Get leaf proofs.
+	proofs, err := s.inclusionProofs(ctx, indexes, in.Epoch)
+	if err != nil {
+		return nil, err
+	}
+	for i, p := range proofs {
+		mutations[i].Proof = p
+	}
+
+	// Fetch log proofs.
+	logRoot, logConsistency, logInclusion, err := s.logProofs(ctx, in.FirstTreeSize, resp.GetMapRoot().MapRevision)
+	if err != nil {
+		return nil, err
+	}
+
+	nextPageToken := ""
+	if len(mutations) == int(in.PageSize) && maxSequence != highestSeq {
```

Comment:
Not sure I'm following, but this case should work as expected. The last page will have few mutations than the request page size and the nextPageToken will be empty. There's a test case for this already. Please check my comment below.